### PR TITLE
@DAWN-391  added sbrk and fixed the function signature of memset

### PR DESCRIPTION
**Pull request by larryk85**
/libraries/chain/wasm_interface.cpp and /libraries/chain/include/eosio/chain/wasm_interface_private.hpp updated.  These will be needed by anyone adding missing api intrinsics.

### CODE REVIEW

File: libraries/chain/include/eosio/chain/wasm_interface_private.hpp [hunk 1]
```diff
@@ -411,20 +411,23 @@ struct intrinsic_invoker_impl<Ret, std::tuple<array_ptr<T>, array_ptr<U>, size_t
...
  * @tparam Inputs - the remaining native parameters to transcribe
  * @tparam Translated - the list of transcribed wasm parameters
  */
-template<>
-struct intrinsic_invoker_impl<void_type, std::tuple<array_ptr<char>, int, size_t>, std::tuple<>> {
-   using next_step = intrinsic_invoker_impl<void_type, std::tuple<>, std::tuple<I32, I32, I32>>;
-   using then_type = void_type(*)(wasm_interface &, array_ptr<char>, int, size_t);
+template<typename Ret>
+struct intrinsic_invoker_impl<Ret, std::tuple<array_ptr<char>, int, size_t>, std::tuple<>> {
+   //using next_step = intrinsic_invoker_impl<void_type, std::tuple<>, std::tuple<I32, I32, I32>>;
```
Comment (wanderingbort) [inline libraries/chain/include/eosio/chain/wasm_interface_private.hpp]: go ahead and clean up the commented old `using` statements

File: libraries/chain/include/eosio/chain/wasm_interface_private.hpp [hunk 2]
```diff
@@ -411,20 +411,23 @@ struct intrinsic_invoker_impl<Ret, std::tuple<array_ptr<T>, array_ptr<U>, size_t
    }
 };
 
+#if 1
```
Comment (wanderingbort) [inline libraries/chain/include/eosio/chain/wasm_interface_private.hpp]: please clean up this preprocessor conditional

File: libraries/chain/wasm_interface.cpp [hunk 3]
```diff
@@ -23,6 +23,7 @@
 #include <mutex>
 #include <thread>
 #include <condition_variable>
+//#include <iostream>
```
Comment (wanderingbort) [inline libraries/chain/wasm_interface.cpp]: no need to keep this around

File: libraries/chain/wasm_interface.cpp [hunk 4]
```diff
@@ -622,11 +623,12 @@ DEFINE_INTRINSIC_FUNCTION0(env,checktime,checktime,none) {
 class context_aware_api {
    public:
       context_aware_api(wasm_interface& wasm)
-      :context(intrinsics_accessor::get_context(wasm).context)
+      :context(intrinsics_accessor::get_context(wasm).context), code(intrinsics_accessor::get_context(wasm).code)
       {}
 
    protected:
-      apply_context& context;
+		wasm_cache::entry& code;
```
Comment (wanderingbort) [inline libraries/chain/wasm_interface.cpp]: make sure you have "tabs" set to spaces

File: libraries/chain/wasm_interface.cpp [hunk 5]
```diff
@@ -819,10 +822,44 @@ class memory_api {
...
+			constexpr uint32_t 	NBPPL2 = IR::numBytesPerPageLog2;
+			constexpr uint32_t max_mem = 1024 * 1024;
+			const auto 			 default_mem = Runtime::getDefaultMemory(code.instance);
+			const uint32_t		 num_pages = Runtime::getMemoryNumPages(default_mem);
+			// limit this min to 32 bit space
+			const uint32_t 	 min_bytes = (num_pages << NBPPL2) > UINT32_MAX ? UINT32_MAX : num_pages << NBPPL2;
+			static uint32_t 	 _num_bytes = min_bytes;
+			const uint32_t 	 prev_num_bytes = _num_bytes;
+
+			if (Runtime::getMemoryNumPages(default_mem) != num_pages)
```
Comment (wanderingbort) [inline libraries/chain/wasm_interface.cpp]: this seems redundant given that we get this value as const.

**Comment by spoonincode** [comment]
I think we need to update eoslib/memory.h for the new memset() prototype too

**Comment by spoonincode** [comment]
@wanderingbort you are right -- the design assumes that you're using malloc() to call sbrk() -- the memset occurs in malloc(). This looks like a massive bug since someone could use sbrk() without malloc(). IMO the WASM runtime needs to be refactored to always ensure you're given zeroed memory. I'm going to have to touch that code significantly when changing the virtual mem allocation strategy so I wouldn't bother fixing any of that logic now.

**Comment by wanderingbort** [comment]
Just waiting on a fix from @larryk85 that will deallocate the pages properly before returning the code to the cache

**Pull request merged**